sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java

```diff
@@ -160,39 +167,104 @@ public FunctionImplementation getForTypes(List<DataType> args) throws IllegalArg
-                    throw new UnsupportedOperationException(
-                        operator + " is not supported on expressions of type " + dataType.getName());
+            assert dataTypes.size() == 2 : "Arithmetic operator must receive two arguments";
+            if (dataTypes.get(0).equals(dataTypes.get(1))) {
+                DataType dataType = dataTypes.get(0);
+                final BinaryScalar<?> scalar;
+                switch (dataType.id()) {
+                    case DoubleType.ID:
+                        scalar = new BinaryScalar<>(doubleFunction, name, DataTypes.DOUBLE, features);
+                        break;
+
+                    case FloatType.ID:
+                        scalar = new BinaryScalar<>(floatFunction, name, DataTypes.FLOAT, features);
+                        break;
+
+                    case ByteType.ID:
+                    case ShortType.ID:
+                    case IntegerType.ID:
+                        scalar = new BinaryScalar<>(integerFunction, name, DataTypes.INTEGER, features);
+                        break;
+
+                    case LongType.ID:
+                    case TimestampType.ID_WITH_TZ:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        scalar = new BinaryScalar<>(longFunction, name, DataTypes.LONG, features);
+                        break;
+
+                    default:
+                        throw new UnsupportedOperationException(
+                            operator + " is not supported on expressions of type " + dataType.getName());
+                }
+                return Scalar.withOperator(scalar, operator);
+            }
+            DataType fst = dataTypes.get(0);
+            DataType snd = dataTypes.get(1);
+
+            if (snd.id() == IntervalType.ID) {
+                switch (fst.id()) {
+                    case TimestampType.ID_WITH_TZ:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        return new IntervalToTimestampScalar(operator);
+                    default:
+                        throw new UnsupportedOperationException("Invalid operator on intervals");
+                }
+            }
+            throw new UnsupportedOperationException("Invalid operation");
+        }
+    }
+
+    private static final class IntervalToTimestampScalar extends Scalar<Long, Object> {
+
+        private final String operator;
+
+        IntervalToTimestampScalar(String operator) {
+            this.operator = operator;
+        }
+
+        @Override
+        public Long evaluate(TransactionContext txnCtx, Input<Object>... args) {
+            if (args.length != 2) {
+                throw new UnsupportedOperationException("Invalid arguments " + Arrays.toString(args));
+            }
+            Long timestamp = (Long) args[0].value();
+            Period period = (Period) args[1].value();
+            if (timestamp != null && period != null) {
+                switch (operator) {
```

Comment:
Would be nice if we could do this distinction when the method is created, instead of doing it on runtime per row.